Task5 rpq research

vladislav-miroshnikov opened the pull request:
# Задача 5. Экспериментальное исследование алгоритмов для регулярных запросов

[Ссылка на ноутбук](https://colab.research.google.com/drive/1MDsMfnojLoxMknZ8-VpeHM-RlrDLU595#scrollTo=iZCGk4bYma1Q)

Задача посвящена анализу производительности различных реализаций алгоритма решения задачи достижимости между всеми парами вершин с регулярными ограничениями через тензорное произведение.

Решение данной задачи оформляется как Python notebook на [Google colab](https://colab.research.google.com/). Для того, чтобы обеспечить возможность проверки, решение делается доступным по ссылке, ссылка добавляется в readme репозитория, в котором выполняются остальные задачи, делается реквест с этим изменением.

Решением является не просто код, но отчёт об экспериментальном исследовании, который должен являться связанным текстом и содержать (как минимум) следующие разделы:
- Постановка задачи
- Выбор и описание исследуемых решений
- Описание набора данных для экспериментов
  - Графы
  - Запросы
- Описание эксперимента
  - Оборудование
  - Что и как замерялось
- Результаты экспериментов
  - Графики, таблицы
- Анализ результатов экспериментов
  - Какой алгоритм и в каких случаях лучше? Почему?

- [x] Используя [pyCuBool](https://pypi.org/project/pycubool/) реализовать алгоритм пересечения двух конечных автоматов через тензорное произведение. Аналогично тому, как это было сделано в предыдущей домашней работе. Лучше сразу предусмотреть общий интерфейс функций, чтобы было проще ставить эксперименты.
- [x] Подключить решение из предыдущей работы.
- [x] Сформировать набор данных.
  - [x] Выбрать все графы из раздела [RDF](https://jetbrains-research.github.io/CFPQ_Data/dataset/RDF.html).
  - [x] Используя функцию из первой домашней работы узнать метки рёбер графов и на основе этой информации сформулировать не менее 4 различных запросов к каждому графу. Требования к запросам:
      - Запросы ко всем графам должны следовать некоторому общему шаблону. Например, если есть графы ```g1``` и ```g2``` с различными наборами меток, то ожидается, что запросы к ним будут выглядеть, например, так:
        - ```g1```:
          - ```(l1 | l2)* l3```
          - ```(l3 | l4)+ l1*```
          - ```l1 l2 l3 (l4|l1)*```
        - ```g2```:
          - ```(m1 | m3)* m2```
          - ```(m1 | m3)+ m2*```
          - ```m1 m2 m3 (m3|m1)*```
      - В запросах должны использоваться все общепринятые конструкции регулярных выражений  (замыкание, конкатенация, альтернатива). То есть хотя бы в одном запросе к каждому графу должна быть каждая из этих конструкций.
- [x] С использованием сформированного набора данный провести сравнение производительности алгоритма, основанного на тензорном произведении, реализованного с использованием двух библиотек
  - sciPy (предыдущая работа)
  - pyCuBool (первый пункт этой работы)
- [x] Оформить результаты экспериментов
- [x] Провести анализ результатов
- [x] Сделать notebook доступным по ссылке

gsvgit committed:
Ещё несколько опечаток в пятом задании.

vladislav-miroshnikov committed:
Added research notebook

vladislav-miroshnikov committed:
Exclude experiments from pre-commit

vladislav-miroshnikov committed:
Added final version of notebook

gsvgit committed:
Задача 6. Преобразование в ОНФХ.

gsvgit committed:
Merge remote-tracking branch 'origin/main' into tasks

gsvgit committed:
Merge pull request #10 from JetBrains-Research/tasks

Задача 6

vladislav-miroshnikov committed:
Merge branch 'main' of https://github.com/JetBrains-Research/formal-lang-course

vadyushkins reviewed (changes requested):
* Проблема №1. "лучший результат из 3 запросов"
  * Непонятно какой именно результат замеров получается
* Проблема №2. "6 знаков после запятой"
  * С тремя замерами Вы не можете оценить даже первый десятичный знак

Остальные комментарии оставил в ноутбуке

> P.S. Так как требуется исправления уже после жесткого дедлайна, то за задачу не более `5.5` баллов

gsvgit committed:
Задача 7. Рекурсивные конечные автоматы.

gsvgit committed:
Уточнение формулировки задачи 6. Раница между НФХ и ОНФХ.

gsvgit committed:
Задача 6. Добавлено определение ОНФХ.

gsvgit committed:
Уточнения постановки задачи 7.

- Указано, что ожидается от минимизации
- Явно указано, что входная граммтика может "появиться откуда угодно".

gsvgit committed:
Переформулировка задачи 7

1) Теперь CFG -> ECFG -> RSM
2) Вынес в отдельный подраздел описание ECFG.

gsvgit committed:
Merge pull request #11 from JetBrains-Research/tasks

Задача 7. Рекурсивные конечные автоматы.

gsvgit committed:
Merge pull request #12 from JetBrains-Research/fix_task_6

Уточнение формулировки задачи 6. Раница между НФХ и ОНФХ.

vladislav-miroshnikov committed:
Merge branch 'main' of https://github.com/JetBrains-Research/formal-lang-course

gsvgit committed:
Задача 8. CYK.

gsvgit committed:
Merge pull request #13 from JetBrains-Research/tasks

Задача 8. CYK.

vladislav-miroshnikov committed:
Merge branch 'main' of https://github.com/JetBrains-Research/formal-lang-course

vladislav-miroshnikov committed:
Added new version of research notebook

vladislav-miroshnikov commented:
> * Проблема №1. "лучший результат из 3 запросов"
>   
>   * Непонятно какой именно результат замеров получается
> * Проблема №2. "6 знаков после запятой"
>   
>   * С тремя замерами Вы не можете оценить даже первый десятичный знак
> 
> Остальные комментарии оставил в ноутбуке
> 
> > P.S. Так как требуется исправления уже после жесткого дедлайна, то за задачу не более `5.5` баллов

Исправил замечания. Во-первых, я заменил "лучший результат из 3 запросов" и пояснил, что имеется в виду. Далее, как вы и говорили, для оценки одного десятичного знака я использовал 10 замеров, а не 3.

vadyushkins reviewed (approved):
Зачтено

vladislav-miroshnikov committed:
Merge branch 'main' into task5-rpq-research
